### ISSUE TITLE
SequenceGenerator won't work without a Fork layer between feedback and recurrent transition

### ISSUE THREAD

**Issue by mjwillson** [bartvm/blocks#589]
I wanted to try a basic RNN language model where the word representation fed back from the previous timestamp (via LookupFeedback) is added into the hidden state without being run through an additional linear layer.

It seems SequenceGenerator really wants to insert a Fork in between the feedback and the hidden state though. Specifying fork=None doesn't work, nor does fork=Identity(), since it expects fork.apply.outputs to exist as input names for the recurrent transition's apply, and Identity's output name of "output" doesn't exist there.

Perhaps Fork is the default because it makes sense for a GatedRecurrent, but with a SimpleRecurrent I was a bit surprised to find that it had inserted this extra layer by default without me asking for it (and the fact that it's hidden behind a Fork with a single output makes it less clear that actually it's going to insert a linear layer).

For now I think I've worked around it by creating my own Identity-like brick to use as the fork, which has a custom output name of "inputs" for its apply method (this is what the input name needs to be for my recurrent layer). Feels a bit weird for the output name of a layer to be inputs -- bit confused by all this plumbing based on string-based input and output names to be honest! But if it works I guess...

**Comment by mjwillson** [comment]
Also just as another note, that a Linear layer following after a LookupFeedback may often be redundant overparameterisation, since LookupTable is effectively a linear layer too if you consider the inputs as being one-hot word vectors, and two linear layers in a row can be redundant depending on the dimensions involved.

For that reason I feel like this extra fork layer should probably be off by default, at least for SimpleRecurrent. Why is it that you need two different hooks (feedback and fork) to substitute in extra transformations between the observation from the previous timestep, and its contribution to the next recurrent transition? Feels like it only really makes sense for things like GatedRecurrent.

(As another note, this was something about the markov chain example was based on GatedRecurrent it wasn't obvious to me that I'd have to find a way of turning this fork behaviour off if I wanted to apply that example with a SimpleRecurrent)

**Comment by rizar** [comment]
You raised a good point: it is indeed not nice that people using `SimpleRecurrent` are forced to linearly transform their embeddings although this is completely unnecessary. Looks like we should make the fork optional as it was at some point before. After I merge #563, would you like to make a PR for that?

> (As an aside it feels odd to me as a newbie for an output name to be "inputs" -- bit confused by all this plumbing based on string-based input and output names to be honest, e.g. how are they scoped, are they generally supposed to match up when you compose bricks?)

Input and output names are typically used by higher level bricks that have to chain bricks with multiple inputs and outputs. Although this is weird that your fork outputs has to be called `inputs`, I can not think of a principally different solution.

**Comment by mjwillson** [comment]
Cool, OK. Yeah PR should be fine once I'm a little bit more familiar with the codebase :)

About naming of inputs / outputs: I guess if these are important for compositionality (e.g. if higher-level bricks require particular input/output names for their children), then it feels like either every brick that wants to be reusable in a general context should make sure that its input/output names are customizable.

Slightly more elegant(?) alternative might be to accept that each brick class has fixed input/output names, but when you compose them using a higher-level brick, you explicitly tell it which outputs to wire up to which inputs (perhaps with defaults if necessary)

**Comment by rizar** [comment]
> I guess if these are important for compositionality

Yes

> (e.g. if higher-level bricks require particular input/output names for their children)

We (almost) do not do this (with the exception that `BaseSequenceGenerator` expects that mask should be named `mask`). We rather require that input and output names of lower bricks match. For this reason if you want to have a brick that provides inputs to other bricks, it ends up having outputs named `inputs`.

> n it feels like either every brick that wants to be reusable in a general context should make sure that its input/output names are customizable.

Not exactly. Some of the bricks (e.g. `SimpleRecurrent`) might have fixed input and output names, they can be interleaved with various adapters which are fully configurable in this respect, e.g. `Fork`.

> Slightly more elegant(?) alternative might be to accept that each brick class has fixed input/output names, but when you compose them using a higher-level brick, you explicitly tell it which outputs to wire up to which inputs (perhaps with defaults if necessary)

That sounds reasonable in general, and I think we might at some point have a general `NamedStack` brick which will arrange a few bricks in stack and wire them according to an explicit specification given in the constructor. In `BaseSequenceGenerato`r` it seems to be a bit overkill though, because `Fork` has fully configurable output names and can be always created in such a way that it wires `transition` and `readout`.

And it just struck me that what you most likely want now is to create a `Fork` with an `Identity` prototype. Would this work for you?

**Comment by mjwillson** [comment]
Ah OK, gotcha, so things like Fork and Merge are containers that you might need to use to compose bricks inside other bricks, as much as they are about splitting or merging signals. And it's assumed in places that you will use them if you need to compose something.

Maybe worth mentioning this a bit more prominently in their docs, or changing some of the API naming for these bricks to something a bit more general and suggestive (e.g. "routing bricks" as in the docs, or adapter bricks or similar) -- at the moment the module name and superclass is "parallel" which makes it a bit non-obvious that these are general adapter bricks not just things that relate specifically to forking and merging and so on.

I can try wrapping an Identity in a Fork, I'll give that a try! Cheers.

**Comment by mjwillson** [comment]
Sorry about the big tide of doc feedback btw, glad to see you guys care about this stuff though!

**Comment by rizar** [comment]
I agree that the file name 'parallel.py' is a bit misleading. Maybe it should actually be called 'routers.py'. What do you think?

Your feedback is great and we do care about docs, just that I am now a bit under pressure to finish my Masters which makes documentation PRs hang in the air for weeks, sorry for that.

**Comment by rizar** [comment]
Hey @mjwillson , did you try `Fork(Identity())`?

**Comment by mjwillson** [comment]
Yep I did, that worked for me, thanks.

**Comment by rizar** [comment]
Okay, then I would say that we should not make `Fork` optional, but rather keep it mandatory for sake of simplicity. I guess adding a `fork_prototype` to the list of optional parameters of `SequenceGenerator` would be helpful, in this case people would less often need to provide their custom fork.
